### PR TITLE
Remove type assertions and ts-ignore from compose function

### PR DESCRIPTION
**Pull request by captain-yossarian**
Remove type assertions and ts-ignore from compose function

### CODE REVIEW

**Comment by drizzer14** [comment]
Hi! Thanks for the contribution, and nice to meet you outside of reddit :)

I've checked your solution to my comment on line `25` of `compose.ts` locally, and it seems to be working fine, so that's a good thing.

If you don't mind, I'll leave some comments so that you can complete your idea in terms of this project. Otherwise, I can step in myself and finish this PR of yours.

File: src/compose.ts [hunk 1]
```diff
@@ -12,33 +14,41 @@ import type { Last } from './types/last'
...
+    Last
+  ]
+  : Functions
 
 /**
  * Applies all of the provided `functions` one-by-one in right-to-left order
  * starting from the `argument`.
  */
-export default function compose<Functions extends Array<(arg: any) => any>> (
+export default function compose<Functions extends Array<Fn>>(
```
Comment (drizzer14) [inline src/compose.ts:44]: `Fn` type here is actually incorrect, as `compose` directly implies that we're working with **unary** functions (accept exactly **one** argument) here, and the `Fn` is a variadic one (accepts an arbitrary amount of arguments).

**Comment by captain-yossarian** [comment]
What I need to do in order to finish this PR?

File: src/compose.ts [hunk 2]
```diff
@@ -12,33 +14,41 @@ import type { Last } from './types/last'
...
+  ...functions: Composition<Functions>
+): (arg: Parameters<Last<Functions>>[0]) => ReturnType<Functions[0]>
+export default function compose<Functions extends Array<Fn>>(
   ...functions: Composition<Functions>
-): (arg: Parameters<Last<Functions>>[0]) => ReturnType<Functions[0]> {
-  return (arg) => {
-    const length = (functions as any[]).length
+) {
+  return (arg: Parameters<Last<Functions>>[0]) => {
```
Comment (drizzer14) [inline src/compose.ts:50]: This type signature is excessive as we've explicitly annotated the `arg` on line 47.
Actually, this style of annotations is seen across the whole library – adding types before actual runtime code to make the declarations clear from the start.

**Review by drizzer14 — CHANGES_REQUESTED** [review]
Your PR is nice in two ways:
1. It solves the problem with `compose` (and `pipe`, if you'll take a look at it).
2. It reminded me to do some work on the repo to make it more clear and easy for people willing to contribute :)

**Comment by drizzer14** [comment]
Oh, and one more thing – `fnts` follows the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) guidelines to allow for automatic versioning and changelog generation. Make sure to write your commit messages according to those so that they fit nicely into the already existing system.

File: src/compose.ts (hunk 2 above)
Comment (captain-yossarian) [reply src/compose.ts]: There is a problem with assignability. `Parameters<Last<Functions>>[0]` is treated as `unknown`

**Comment by drizzer14** [comment]
Hey @captain-yossarian!

I had some time to think more about the changes awaiting for review and decided to jump in and clean up the code a bit.

I've basically fixed my own comments above and tried to follow your idea a bit further into this final implementation.

Please, take a look, and I'll merge this if you're ok with the final result.

**Review by drizzer14 — COMMENTED** [review]
Hey @captain-yossarian!

I had some time to think more about the changes awaiting for review and decided to jump in and clean up the code a bit.

I've basically fixed my own comments above and tried to follow your idea a bit further into this implementation.

Please, take a look, and I'll merge this if you're ok with the final result.

**Comment by captain-yossarian** [comment]
Hi @drizzer14 your changes are good. Thank you for asking. Looking forward to contribute more into your project

**Review by drizzer14 — APPROVED** [review]
Thank you again for the contribution, hope you find this interesting to use, too :)

Дякую!

**Comment by captain-yossarian** [comment]
@drizzer14 немає за що!

**Pull request merged**